### ISSUE TITLE
telegraf does not respect http(s)_proxy env variables

### ISSUE THREAD

**Issue by rhruiz** [influxdata/telegraf#1588]
## Bug report

The influx http client (v2) that telegraf uses does not uses the defined http_proxy, https_proxy, no_proxy environment variables and always tries to connect directly to the endpoint configured in telegraf.conf
### Relevant telegraf.conf:

None
### System info:

Telegraf - Version 0.10.4.1
Linux valar0003 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt20-1+deb8u4 (2016-02-29) x86_64 GNU/Linux
Debian 8.3
### Steps to reproduce:
1. configure your influxdb http endpoint in telegraf.conf that is directly accessible by the telegraf machine
2. set the appropriate proxy environment variable (http_proxy or https_proxy)
3. check that metrics are not pushed to influxdb
### Expected behavior:

Connections to influxdb are proxyed
### Actual behavior:

Connections are either refused or blocked pending on your firewall configuration and the following is logged:

016/08/05 14:55:54 Database creation failed: Get http://influxdb.****************/query?db=&q=CREATE+DATABASE+IF+NOT+EXISTS+%22*********%22: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
2016/08/05 14:55:54 Starting Telegraf (version 1.0.0-beta1)

**Comment by sparrc** [comment]
can you provide a link to documentation of the environment variable? is it an issue that should be opened on the influxdb repo?

**Comment by rhruiz** [comment]
The golang http itself provides some helper functions to use those variables: https://golang.org/pkg/net/http/#ProxyFromEnvironment

And I open this in the influxdb if it makes more sense

Thanks

**Comment by rhruiz** [comment]
I have created a related issue on the influx repo.

Closing this one. Thanks.

**Comment by rhruiz** [comment]
Reopening as the comment on https://github.com/influxdata/influxdb/issues/7157 says:

"Telegraf issues should be created in influxdata/telegraf rather than here."

**Comment by arcidodo** [comment]
i need this to! is there a workaround?

**Comment by dougtweedy** [comment]
Any update/workaround on this?

**Comment by csaurav** [comment]
Anywork around for this ? I am stuck here too

**Comment by loque1** [comment]
I am facing the same issue. Most corporate environments do not allow unfettered access to the internet from any server. They are tightly controlled so egress out is through a single source i.e. a web proxy.  From the doc's 

DefaultTransport is the default implementation of Transport and is used by DefaultClient. It establishes network connections as needed and caches them for reuse by subsequent calls. It uses HTTP proxies as directed by the $HTTP_PROXY and $NO_PROXY (or $http_proxy and $no_proxy) environment variables.

https://golang.org/pkg/net/http/#ProxyFromEnvironment

The datadog output uses the following from your code 

func (d *Datadog) Connect() error {
	if d.Apikey == "" {
		return fmt.Errorf("apikey is a required field for datadog output")
	}
	d.client = &http.Client{
		Timeout: d.Timeout.Duration,
	}
	return nil
}


func (d *Datadog) Connect() error {
	if d.Apikey == "" {
		return fmt.Errorf("apikey is a required field for datadog output")
	}
       if *proxy != "" {
                proxyURL, err := url.Parse(*proxy)
                if err != nil {
                        panic("Error parsing proxy URL")
                }
                transport := http.Transport{
                        Proxy:           http.ProxyURL(proxyURL)
                }
                client = http.Client{
                        Transport: &transport,
                        Timeout:   time.Duration(time.Millisecond * timeout),
                }

        } else {
                client = http.Client{
                    Timeout: d.Timeout.Duration,
               }
        }
	return nil
}


I think at this point as you overide the default client you need to set the transport to be proxy and provide the proxy url which should be picked up from the environment variables one of HTTP_PROXY, HTTPS_PROXY,http_proxy or https_proxy. 

Does this make sense ? I am not a coder so I suggest somone look over what I am saying

**Comment by loque1** [comment]
Ok My mistake I think it is using the proxy.  setting the following 
HTTPS_PROXY=http://host:port
HTTP_PROXY=http://host:port
https_proxy=http://host:port
http_proxy=http://host:port
causes the proxy to be used for the datadog plugin atleast

**Comment by Integrative** [comment]
Doesn't work for me, InfluxDB output does not respect environment proxy settings.

declare -x http_proxy="http://172.21.8.35:3128"
declare -x https_proxy="http://172.21.8.35:3128"

telegraf /etc/telegraf/telegraf.conf --debug

Post http://xxx:8086/query?db=&q=CREATE+DATABASE+%22cem%22: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
(version 1.2.1)

**Comment by csaurav** [comment]
I have tried with both the http_proxy and https_proxy. But it does not work.

On Mon, Mar 13, 2017, 09:02 Integrative.it <notifications@github.com> wrote:

Doesn't work for me, InfluxDB output does not respect environment proxy
settings.

declare -x http_proxy="http://172.21.8.35:3128"
declare -x https_proxy="http://172.21.8.35:3128"

telegraf /etc/telegraf/telegraf.conf --debug

Post http://xxx:8086/query?db=&q=CREATE+DATABASE+%22cem%22: net/http:
request canceled while waiting for connection (Client.Timeout exceeded
while awaiting headers)
(version 1.2.1)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/influxdata/telegraf/issues/1588#issuecomment-286008013>,
or mute the thread

**Comment by oliven31** [comment]
Any update/workaround on this?

**Comment by jcmcken** [comment]
Any chance this gets into a 1.4.x release? This bug/missing feature renders the ``http_response`` input pretty useless for anything but intranet resources.

**Comment by danielnelson** [comment]
@jcmcken The fix for this issue only added support to the InfluxDB and other output plugins, I opened a new issue for http_response: https://github.com/influxdata/telegraf/issues/3297
